edgedb/retry.py

```diff
@@ -10,16 +10,25 @@ def __init__(self, retry, owner, iteration):
         super().__init__(owner, retry._options.transaction_options)
         self.__retry = retry
         self.__iteration = iteration
+        self.__started = False
 
     async def start(self):
         if not self._managed:
             raise errors.InterfaceError(
                 "Only managed retriable transactions are supported. "
                 "Use `async with transaction:`"
             )
-        await self._start(single_connect=self.__iteration != 0)
+
+    async def _ensure_transaction(self):
+        await self.start()
+        if not self.__started:
+            self.__started = True
+            await self._start(single_connect=self.__iteration != 0)
```

Comment:
Should `self.__started` be set in a `try..finally`?

Reply:
Or just after the `await` line?

Reply:
Like this?

```python
    async def _ensure_transaction(self):
        await self.start()  # only managed transaction check
        if not self.__started:
            try:
                await self._start(single_connect=self.__iteration != 0)
            finally:
                self.__started = True
```

And the reason is - I'm guessing wildly - to allow running `self._start()` concurrently and one of them could raise a more reasonable error? That's actually a good point to have better error messages for wrong concurrency - but `try..finally` won't work, I'll add an `else` branch to check the transaction status and raise friendly error messages.

Reply:
> And the reason is - I'm guessing wildly - to allow running self._start() concurrently

Ah, sure, keep it as is and disregard my earlier suggestion to move `started = True` after await.